webapi/src/main/scala/org/knora/webapi/messages/admin/responder/projectsmessages/ProjectsMessagesADM.scala

```diff
@@ -505,130 +506,61 @@ case class ProjectADM(
   }
-}
 
-/**
- * Represents the project's identifier. It can be an IRI, shortcode or shortname.
- *
- * @param maybeIri       the project's IRI.
- * @param maybeShortname the project's shortname.
- * @param maybeShortcode the project's shortcode
- */
-class ProjectIdentifierADM private (
-  maybeIri: Option[IRI] = None,
-  maybeShortname: Option[String] = None,
-  maybeShortcode: Option[String] = None
-) {
-
-  // squash and return value.
-  val value: String = List(
-    maybeIri,
-    maybeShortname,
-    maybeShortcode
-  ).flatten.head
-
-  def hasType: ProjectIdentifierType =
-    if (maybeIri.isDefined) {
-      ProjectIdentifierType.IRI
-    } else if (maybeShortcode.isDefined) {
-      ProjectIdentifierType.SHORTCODE
-    } else {
-      ProjectIdentifierType.SHORTNAME
-    }
+  case class Shortcode(value: String) extends ProjectIdentifierADM {
+    def apply(value: String): ShortCode = ShortCode.make(value).fold(e => throw e.head, v => v)
+  }
 
-  /**
-   * Tries to return the value as an IRI.
-   */
-  def toIri: IRI =
-    maybeIri.getOrElse(
-      throw DataConversionException(s"Identifier $value is not of the required 'ProjectIdentifierType.IRI' type.")
-    )
+  case class Shortname(value: String) extends ProjectIdentifierADM {
+    def apply(value: String): ShortName = ShortName.make(value).fold(e => throw e.head, v => v)
+  }
 
-  /**
-   * Returns an optional value of the identifier.
-   */
-  def toIriOption: Option[IRI] =
-    maybeIri
+  case class Uuid(value: String) extends ProjectIdentifierADM
 
-  /**
-   * Tries to return the value as an SHORTNAME.
-   */
-  def toShortname: String =
-    maybeShortname.getOrElse(
-      throw DataConversionException(s"Identifier $value is not of the required 'ProjectIdentifierType.SHORTNAME' type.")
-    )
+  def asIriOption(id: ProjectIdentifierADM): Option[String] =
+    id match {
+      case Iri(value) => Some(value)
+      case _          => None
+    }
 
-  /**
-   * Returns an optional value of the identifier.
-   */
-  def toShortnameOption: Option[String] =
-    maybeShortname
+  def asShortcodeOption(id: ProjectIdentifierADM): Option[String] =
+    id match {
+      case Shortcode(value) => Some(value)
+      case _                => None
+    }
 
-  /**
-   * Tries to return the value as an SHORTCODE.
-   */
-  def toShortcode: String =
-    maybeShortcode.getOrElse(
-      throw DataConversionException(s"Identifier $value is not of the required 'ProjectIdentifierType.SHORTCODE' type.")
-    )
+  def asShortnameOption(id: ProjectIdentifierADM): Option[String] =
+    id match {
+      case Shortname(value) => Some(value)
+      case _                => None
+    }
 
-  /**
-   * Returns an optional value of the identifier.
-   */
-  def toShortcodeOption: Option[String] =
-    maybeShortcode
+  def asUuidOption(id: ProjectIdentifierADM): Option[String] =
+    id match {
+      case Uuid(value) => Some(s"http://rdfh.ch/projects/${value}")
```

Comment:
This case does not return  `Uuid` as promised by the function. This returns something which looks like a `String` representation of a `ProjectIri`, isn't it?

Reply:
You are right. This is kind of fake case class which as stated in the PR description reuses IRI methods. I think this would stay like that for now and should be changed once all IRI types will be transformed to new shape with UUID instead of shortcode.